File: ios/GRIDDOWN/Services/CheckIn.swift
import SwiftUI

enum CheckInState: String, Sendable {
    case never, current, dueSoon, overdue

    var label: String {
        switch self {
        case .never: "NO CHECK-IN"
        case .current: "CURRENT"
        case .dueSoon: "DUE SOON"
        case .overdue: "OVERDUE"
        }
    }

    var color: Color {
        switch self {
        case .never: Theme.textMuted
        case .current: Theme.statusGreen
        case .dueSoon: Theme.statusAmber
        case .overdue: Theme.statusRed
        }
    }
}

/// Check-in cadence logic: a member is "dueSoon" once 75% of the interval has
/// elapsed and "overdue" once the full interval passes without a check-in.
enum CheckIn {
    static let intervalOptions = [2, 4, 6, 8, 12, 24]

    static func state(for member: GroupMember, intervalHours: Int, now: Date = Date()) -> CheckInState {
        guard let raw = member.lastCheckInAt, let last = parse(raw) else { return .never }
        let elapsed = now.timeIntervalSince(last)
        let interval = TimeInterval(intervalHours) * 3600
        if elapsed >= interval { return .overdue }
        if elapsed >= interval * 0.75 { return .dueSoon }
        return .current
    }

    static func timeSince(_ iso: String?, now: Date = Date()) -> String {
        guard let iso, let date = parse(iso) else { return "Never" }
        let seconds = now.timeIntervalSince(date)
        if seconds < 0 { return "Just now" }
        let minutes = Int(seconds / 60)
        if minutes < 1 { return "Just now" }
        if minutes < 60 { return "\(minutes)m ago" }
        let hours = minutes / 60
        if hours < 24 { return "\(hours)h ago" }
        return "\(hours / 24)d ago"
    }

    /// Returns how long a member is past their expected check-in, or nil if not overdue.
    static func overdueBy(_ member: GroupMember, intervalHours: Int, now: Date = Date()) -> String? {
        guard let raw = member.lastCheckInAt, let last = parse(raw) else { return nil }
        let overdue = now.timeIntervalSince(last) - TimeInterval(intervalHours) * 3600
        if overdue <= 0 { return nil }
        let minutes = Int(overdue / 60)
        if minutes < 60 { return "\(minutes)m overdue" }
        let hours = minutes / 60
        if hours < 24 { return "\(hours)h overdue" }
        return "\(hours / 24)d overdue"
    }

    private static func parse(_ iso: String) -> Date? {
        ISO8601DateFormatter().date(from: iso)
    }
}

File: expo/utils/checkin.ts
import { GroupMember } from '@/types';
import Colors from '@/constants/colors';

export type CheckInState = 'never' | 'current' | 'dueSoon' | 'overdue';

export const DEFAULT_CHECK_IN_INTERVAL_HOURS = 6;
export const CHECK_IN_INTERVAL_OPTIONS = [2, 4, 6, 8, 12, 24];

/**
 * Computes the check-in state for a member given the group's check-in cadence.
 * A member is "dueSoon" once 75% of the interval has elapsed, and "overdue"
 * once the full interval has passed without a check-in.
 */
export function getCheckInState(
  member: GroupMember,
  intervalHours: number,
  now: Date = new Date()
): CheckInState {
  if (!member.lastCheckInAt) return 'never';
  const last = new Date(member.lastCheckInAt);
  if (Number.isNaN(last.getTime())) return 'never';
  const elapsedMs = now.getTime() - last.getTime();
  const intervalMs = intervalHours * 60 * 60 * 1000;
  if (elapsedMs >= intervalMs) return 'overdue';
  if (elapsedMs >= intervalMs * 0.75) return 'dueSoon';
  return 'current';
}

export function checkInStatusLabel(state: CheckInState): string {
  switch (state) {
    case 'never': return 'NO CHECK-IN';
    case 'current': return 'CURRENT';
    case 'dueSoon': return 'DUE SOON';
    case 'overdue': return 'OVERDUE';
  }
}

export function checkInStatusColor(state: CheckInState): string {
  switch (state) {
    case 'never': return Colors.textMuted;
    case 'current': return Colors.statusGreen;
    case 'dueSoon': return Colors.statusAmber;
    case 'overdue': return Colors.statusRed;
  }
}

/**
 * Formats an ISO timestamp as a short relative string, e.g. "25m ago", "3h ago", "2d ago".
 */
export function formatTimeSince(iso?: string, now: Date = new Date()): string {
  if (!iso) return 'Never';
  const date = new Date(iso);
  if (Number.isNaN(date.getTime())) return 'Never';
  const ms = now.getTime() - date.getTime();
  if (ms < 0) return 'Just now';
  const minutes = Math.floor(ms / 60000);
  if (minutes < 1) return 'Just now';
  if (minutes < 60) return `${minutes}m ago`;
  const hours = Math.floor(minutes / 60);
  if (hours < 24) return `${hours}h ago`;
  const days = Math.floor(hours / 24);
  return `${days}d ago`;
}

/**
 * Returns how long a member is past their expected check-in, or null if not overdue.
 */
export function overdueBy(member: GroupMember, intervalHours: number, now: Date = new Date()): string | null {
  if (!member.lastCheckInAt) return null;
  const last = new Date(member.lastCheckInAt);
  if (Number.isNaN(last.getTime())) return null;
  const overdueMs = now.getTime() - last.getTime() - intervalHours * 60 * 60 * 1000;
  if (overdueMs <= 0) return null;
  const minutes = Math.floor(overdueMs / 60000);
  if (minutes < 60) return `${minutes}m overdue`;
  const hours = Math.floor(minutes / 60);
  if (hours < 24) return `${hours}h overdue`;
  return `${Math.floor(hours / 24)}d overdue`;
}
